### DISCUSSION TITLE
Find correspondences between a set of 3D gaussian distributions (quadrics or ellipsoids) and a set of 2d gaussian distributions (conics or ellipses)

### DISCUSSION THREAD

**Discussion by ttsesm** [PythonOT/POT#599] [Q&A]
Hello,

I don't have a formal background in OT, therefore with this issue I am also trying to understand a bit better whether OT could be an option for my task or not.

So briefly, I am having the following problem. Imagine that I have set of 3D gaussian distributions, i.e quadrics or Ellipsoids, in the 3d space and a set of 2D gaussian distributions, i.e. conics or ellipses, projected on an plane. This number of 3D/2D gaussians in each set could be couple of thousands. Moreover, for each of the sets I have all the information that describe these two sets as gaussian distributions respectively. Thus, the mean, μ, and the covariance matrices Σ.

My question is, based on this information is there an approach that I could possibly obtain some good correspondences so that I can match an ellipsoid with an ellipse. Each distribution in each of the sets can be considered as individual distributions or combined. Moreover, it is not necessary that all distributions in one set match with all distributions on the other set. It can be that only few ellipsoids match with some of the ellipses in the other set. I am trying to figure out whether geometrically or statistically is possible to get something based on OT.

![Picture11](https://github.com/PythonOT/POT/assets/10189018/f06789dc-ed14-49a3-92c0-fabbc0f91e29)

The inspiration comes from these works: [Generalized Wasserstein barycenters between
probability measures living on different subspaces](https://hal.science/hal-03231278/document) and this thesis [Generalised Wasserstein Barycentres
](https://eloitanguy.github.io/assets/pdf/GWB_report.pdf) where some of the authors (@eloitanguy, @rflamary) from what I've noticed are also contributors of the POT library. The idea as described there is that based on some probability measures, marginals or projections (≥2) using OT it possible to recover the 3d probability or the reconstructed generalized barycenter which apparently describes the geometry of the distribution. In my case though I have only one projection, which of course this projection contains multiple 2D gaussians.

To my understanding the Blind Generalised Wasserstein Barycentre Problem could be an approach to follow but I am not quite sure since the whole concept is totally new to me. Thus,  any feedback would be appreciated.

**Comment by eloitanguy** [comment]
Hi,

With BGWB (which is not implemented in POT), you need discrete distributions. A possible avenue would be to sample from your 3D  and 2D gaussians, compute a projection P: R^3 -> R^2 which maps your samples, as well as a transport plan $\pi$ which will tell you which 3d sample matches which 2d sample.

The minimisation problem in question would be $\min_{P}W_2^2(P$# $\mu, \nu)$, where $\mu$ is your 3D samples and $\nu$ is your 2D samples. You could solve this using BCD or SGD (similarly to  [Generalised Wasserstein Barycentres](https://eloitanguy.github.io/assets/pdf/GWB_report.pdf) ), but this could yield poor results numerically.

To use Gaussians and not samples, you could see each Gaussian as a point on the Bures-Wasserstein manifold. To compute an assignment between your 3D gaussians and 2D gaussians, you could then consider the Gromov-Wasserstein distance, which can correspond elements of different metric spaces.

In general, your problem also makes me think of Domain Adaptation, which has been tackled with OT successfully.

Thanks for the interesting question, I don't think we can give an immediate answer with accessible tools :D
Let us know how your thoughts go along!

**Reply by ttsesm** [reply]
Hi Eloi,

Thanks a lot for your feedback and your time, appreciated. 

So, from what I've understood your suggestion is to work directly on gaussians instead of the samples (btw, with the term samples what do you mean? could you please elaborate a bit more, I didn't really get that clearly... do you mean marginals/projections or something else?).

Based on the gaussians one approach that I had already in my mind was to test the [following example](https://pythonot.github.io/auto_examples/gromov/plot_gromov.html) Gromov-Wasserstein distance from the POT examples pool. If I understand correctly this is what you have in mind, where I guess I consider the mean, μ, of each guassian distribution from my two sets as a point on the Bures-Wasserstein manifold. Would that be a good initiative to start experimenting?

Also why do you mention that my problem could be seen as a domain adaptation task, I would be interested to hear your thought on that.

**Reply by eloitanguy** [reply]
By samples, i mean samples from the Gaussian distributions.

The Gromov-Wasserstein example you mention could be a good starting point: for the 3D dataset, you compute the distance matrix between the Gaussians $g_i = (m_i, \Sigma_i)$ using the Bures-Wasserstein distance (you need the mean and the covariance matrix! your previous message seemed to only mention the mean), this yields a distance matrix $C_1$. For the 2D dataset, you do the same with the 2D Gaussians, yielding a distance matrix $C_2$. Then, you compute (with [ot.gromov.BAPG_gromov_wasserstein](https://pythonot.github.io/gen_modules/ot.gromov.html#ot.gromov.BAPG_gromov_wasserstein) perhaps) the Gromov-Wasserstein plan $T$, whose entry $i,j$ matches the 3D Gaussian $i$ with the 2D Gaussian $j$.

This idea is relatively naive, since it does not really take into account that there is a projection involved, and that perhaps only *some* of the gaussians can actually be matched.

Regarding domain adaptation, i am not an expert, but there seems to be a common theme of identifying objects (samples, in DA) in different spaces.

Hope this helps!

**Reply by ttsesm** [reply]
Thanks a lot Eloi, let me set up a small example with your proposals and see what I can get. Then we can continue the discussion also based on some practical outputs. 

Taking into account the projection map, _P_, as well as the fact that some of the gaussians might only be need to be matched will be needed. In first place though, I will try on a toy example where all 3d gaussians have a 2d gaussian correspondence and see how this performs.

**Reply by ttsesm** [reply]
Hi @eloitanguy,

Ok, I got some time to play a bit with the solution that you suggested. I will try to elaborate as much as possible with code as well so that we are in the same page.

I have the following input, ten ellipsoids and ten ellipses corresponding one to one. For ellipsoids:

Covariances
```
Cs = torch.tensor([[[ 0.00055704,  0.0002181 ,  0.00020697],
        [ 0.0002181 ,  0.00026268,  0.00017932],
        [ 0.00020697,  0.00017932,  0.00013152]],

       [[ 0.00029736,  0.00036497, -0.00025849],
        [ 0.00036497,  0.0005222 , -0.00034746],
        [-0.00025849, -0.00034746,  0.00028704]],

       [[ 0.00013889, -0.00002427,  0.00009434],
        [-0.00002427,  0.000106  , -0.00010998],
        [ 0.00009434, -0.00010998,  0.00020017]],

       [[ 0.00013997,  0.0000322 , -0.00003177],
        [ 0.0000322 ,  0.00013866,  0.0000159 ],
        [-0.00003177,  0.0000159 ,  0.00008226]],

       [[ 0.00015876,  0.00015987,  0.00013226],
        [ 0.00015987,  0.00046588, -0.0000928 ],
        [ 0.00013226, -0.0000928 ,  0.00045325]],

       [[ 0.00014214, -0.00009106,  0.00025811],
        [-0.00009106,  0.00042213,  0.00002785],
        [ 0.00025811,  0.00002785,  0.00058036]],

       [[ 0.00003409, -0.00000352,  0.00000672],
        [-0.00000352,  0.00000058, -0.00000025],
        [ 0.00000672, -0.00000025,  0.00000226]],

       [[ 0.00025886, -0.00007174, -0.00000447],
        [-0.00007174,  0.00020796,  0.00000071],
        [-0.00000447,  0.00000071,  0.00017473]],

       [[ 0.00000316, -0.00008337, -0.00002749],
        [-0.00008337,  0.00292826,  0.00008949],
        [-0.00002749,  0.00008949,  0.00106939]],

       [[ 0.00080745, -0.00143929, -0.00031194],
        [-0.00143929,  0.00750104,  0.00179461],
        [-0.00031194,  0.00179461,  0.0016364 ]]], dtype=torch.float32)
```

Means
```
Ms = torch.tensor([[ 0.17411666,  0.5035763 ,  0.42941484],
       [ 0.5251834 , -0.27657598, -0.18892948],
       [-0.5073861 ,  0.46957424, -0.90752286],
       [ 0.56422734, -0.4835634 , -0.91173404],
       [-0.5370846 ,  0.22988465, -0.2466668 ],
       [ 0.04097709,  0.48167574, -0.23595962],
       [ 0.49995327, -0.5364788 , -0.3129182 ],
       [ 0.4078352 ,  0.46962914, -0.3338474 ],
       [ 0.5331596 , -0.24272342, -0.30946448],
       [-0.57427835, -0.20426048, -0.28931147]], dtype=torch.float32)
```

and for ellipses:
Covariances
```
ct = torch.tensor([[[0.00049392, 0.00018833],
        [0.00018833, 0.00007765]],
       [[0.00037171, 0.00021303],
        [0.00021303, 0.00012497]],
       [[0.00004997, 0.00002954],
        [0.00002954, 0.00012977]],
       [[0.00005111, 0.00000863],
        [0.00000863, 0.00002055]],
       [[0.00022878, 0.00014253],
        [0.00014253, 0.00023158]],
       [[0.00010538, 0.00007804],
        [0.00007804, 0.00030141]],
       [[0.00001159, 0.00000829],
        [0.00000829, 0.00000615]],
       [[0.00007754, 0.00002211],
        [0.00002211, 0.00011025]],
       [[0.00070302, 0.00060664],
        [0.00060664, 0.00052477]],
       [[0.00109267, 0.00122162],
        [0.00122162, 0.00137384]]], dtype=torch.float32)
```
Means
```
mt = torch.tensor([[-0.34575066, -0.05927338],
       [-0.16067392, -0.15656751],
       [ 0.0884882 ,  0.7348885 ],
       [-0.09218572,  0.05595953],
       [ 0.21521865,  0.38351792],
       [-0.22240582,  0.30064186],
       [-0.0520223 , -0.18196055],
       [-0.4030493 ,  0.22060394],
       [-0.17406262, -0.09412329],
       [ 0.3939966 ,  0.21849754]], dtype=torch.float32)
```

Then I've tried to follow your suggestions, thus I have the following code snippet:
```
 W_psoid = torch.zeros([10, 10], dtype=torch.float32)
  W_pse = torch.zeros([10, 10], dtype=torch.float32)

  for i in range(0, 10):
      for j in range(0, 10):
          w = ot.gaussian.bures_wasserstein_distance(Ms[i], Mt[j], Cs[i], Ct[j])
          W_psoid[i,j] = w

          w = ot.gaussian.bures_wasserstein_distance(ms[i], mt[j], cs[i], ct[j])
          W_pse[i,j] = w
```
where `W_psoid` and `W_pse` look good based on the following outputs:

W_psoid:
![image](https://github.com/PythonOT/POT/assets/10189018/85efd644-3452-4da1-bb60-778097219d4c)
W_pse:
![image](https://github.com/PythonOT/POT/assets/10189018/5a1dbb25-585f-4121-8d2a-ed41f2db4d79)

When I try to compute the Gromov-Wasserstein plan though, I am getting the following output which to me looks not good or I am miss-interpreting something:

`ot.gromov.BAPG_gromov_wasserstein(W_psoid, W_pse)`
![image](https://github.com/PythonOT/POT/assets/10189018/dfa084a1-9e3f-42fb-a9bf-ae255ae99b34)

or `ot.gromov.gromov_wasserstein(W_psoid, W_pse)`
![image](https://github.com/PythonOT/POT/assets/10189018/bbe6e302-f67a-4160-905d-dad4722051aa)

to my understanding I should be getting the low values on the diagonal but as you can see this is not the case.

Thus, do you have any ideas why this is happening and any possible workaround.

Thanks.

**Reply by eloitanguy** [reply]
Hi,
Your code with `ot.gromov.gromov_wasserstein(W_psoid, W_pse)` finds a permutation matrix matching your ellipsoids to your ellipses and looks good to me. What is the issue? Why do you expect "low values on the diagonal"? I agree that the output of `ot.gromov.BAPG_gromov_wasserstein(W_psoid, W_pse)` is not satisfactory (it doesn't pass the eyeball test for the marginal constraints, maybe it didn't converge properly? Check the logs if you want this method specifically), but your second output with `ot.gromov.gromov_wasserstein(W_psoid, W_pse)` looks perfectly reasonable, matching each ellipsoid with exactly one ellipsis in this case. Perhaps for your future experiments, build a test set where you can predict a specific output, and check whether the method suits your goals.

**Reply by ttsesm** [reply]
I am saying that the low values should be on the diagonal because this is the correct matching. Meaning that ellipsoid in the position 0 corresponds to the ellipse in position 0, then 1 to 1, 2 to 2, ... so actually this is the test :-)

Thus, I already know that this is my ground truth, but as you can see the output is completely wrong. Only two, have been matched correctly, i.e. the 2 and 6.

**Reply by eloitanguy** [reply]
Understood, so by "low values" you meant the nonzero entries in the case of the second transport plan. It would seem that applying GW directly isn't satisfactory then, but further experimentation could be warranted.
Another simple idea would be to find the 3D -> 2D mapping (assuming it's linear, i don't know if this is the case in your setup), by minimising $\mathrm{W}_2^2(P\\#\mu, \nu)$, where $\mu$ is your 3D Gaussians, $\nu$ is your 2D Gaussians, and $P$ is a linear map $\mathbb{R}^3 \longrightarrow \mathbb{R}^2$, and you take the Bure-Wasserstein distance to build your cost matrix. 
I believe you'll have to resort to optimisation problems like these, which you can solve using tools from PythonOT. Here, gradient descent on $P$ should give you something with some tinkering (note that there are probably numerous local and even global optima in general).

**Reply by ttsesm** [reply]
Yes, you are right my bad. By "low values" I meant the nonzero ones in the second transport plan.

In principle my mapping, `P`, should be indeed linear. Is there any example showcasing such a minimization mapping? Is [this example](https://pythonot.github.io/auto_examples/domain-adaptation/plot_otda_linear_mapping.html#sphx-glr-auto-examples-domain-adaptation-plot-otda-linear-mapping-py) correct?

Yes, I am expecting to have local and global optima. I hope with some tinkering I can find a more generic solution. Also this, mapping will it work in case of noise? where not all the ellipsoids correspond to ellipses and that only some of the gaussians can actually be matched.

**Reply by eloitanguy** [reply]
The example you cite deals with a different scenario which does not apply here, it boils down to finding a linear map matching two measures on (possibly different) Euclidean spaces. In you case, you want to match ellispoids, which are points on the Bures-Wasserstein space.
The idea that i suggest requires you to code it: back-propagation of the gradient with respect to $P$ of the 2-Wasserstein distance (`ot.emd2`) between the 2D-projected gaussians $P\\#\mu$ and the 2D gaussians $\mu$, using the 2D Bures-Wasserstein distance to build your cost matrix (`ot.gaussian.bures_wasserstein_distance` at each entry).

**Reply by ttsesm** [reply]
I see, then what about the [ot.gaussian.bures_wasserstein_mapping()](https://pythonot.github.io/gen_modules/ot.gaussian.html#ot.gaussian.bures_wasserstein_mapping) can't I directly use this function to estimate the mapping :thinking:

**Reply by eloitanguy** [reply]
[ot.gaussian.bures_wasserstein_mapping()](https://pythonot.github.io/gen_modules/ot.gaussian.html#ot.gaussian.bures_wasserstein_mapping) computes the linear map which maps a Gaussian to another. This is not the same problem.

**Reply by ttsesm** [reply]
ok got it, I will try to test the 3D->2D mapping with the `ot.emd2`. 

You also mentioned that, for applying GW directly further experimentation could be possibly applied. Could you elaborate a bit what other approaches I could possibly test there. For example I've tried to play with the function parameters but it didn't really change much. Then is there anything else that could worth trying.

**Reply by eloitanguy** [reply]
> You also mentioned that, for applying GW directly further experimentation could be possibly applied. Could you elaborate a bit what other approaches I could possibly test there. For example I've tried to play with the function parameters but it didn't really change much. Then is there anything else that could worth trying.

No specific ideas right now, and no time to spend on it, sorry. I feel like I have gone above and beyond to give you some starting points on the subject, and this is the most I can reasonably do.
